### PR TITLE
Deprecate remaining VI methods

### PR DESCRIPTION
**Pull request by purna135**
Addressing #5581

Replaced remaining NotImplementedInference methods with deprecation warning.

### CODE REVIEW

**Comment by purna135** [comment]
I need some help with the next step; is adding a deprecation warning the right way to go?

**Comment by ghost** [comment]
I think the idea was to just remove them (even though the title in the issue says "deprecate"). The methods are not implemented in v4, so there is really nothing to deprecate.

**Comment by purna135** [comment]
Could someone please help me remove the Normalizing flows?
I'm not sure if I should remove the entire NormalizingFlowGroup class or just a few lines of code.

https://github.com/pymc-devs/pymc/blob/ef5f91b41d733c64cf3ea785ec2760caae5d1bd4/pymc/variational/approximations.py#L362-L364

**Comment by canyon289** [comment]
@ferrine would you like to take a look here?

File: pymc/variational/approximations.py [hunk 1]
```diff
@@ -270,14 +272,14 @@ def _new_initial(self, size, deterministic, more_replacements=None):
                 at.repeat(self.mean.reshape((1, -1)), size, -1),
                 self.histogram[self.randidx(size)],
             )
+        elif deterministic:
+            warnings.warn(
+                "Deterministic sampling from a Histogram is deprecated and will be removed soon.",
```
Comment (ferrine) [inline pymc/variational/approximations.py]: Is it indeed broken or it was a false alarm?

Comment (ferrine) [reply pymc/variational/approximations.py]: We can enable the test for that and check

File: pymc/variational/opvi.py [hunk 2]
```diff
@@ -957,9 +957,9 @@ def __init_group__(self, group):
         if not group:
             raise GroupError("Got empty group")
         if self.local:
```
Comment (ferrine) [inline pymc/variational/opvi.py]: I think the context of the issue was to entirely delete the code and `local` kwarg, feel free to just delete large portions of the code.

File: pymc/variational/opvi.py [hunk 3]
```diff
@@ -957,9 +957,9 @@ def __init_group__(self, group):
...
-            raise NotImplementedInference("Local inferene aka AEVB is not supported in v4")
+            warnings.warn("Local inferene aka AEVB is deprecated.", DeprecationWarning)
         if self.batched:
```
Comment (ferrine) [inline pymc/variational/opvi.py]: Same for `batched` kwarg and all its mentions. Just delete without hesitation

**Review by ferrine — COMMENTED** [review]
As for normalizing flows I expect something like `rm pymc/variational/flows.py` as a good first step

**Comment by fonnesbeck** [comment]
Is this enough to go on @purna135 ?

**Comment by purna135** [comment]
I'll work on these suggestions and let you know if I need anything else.
Thank you very much.

**Comment by michaelosthege** [comment]
Hi @purna135, this PR is one of the final steps towards `v4.0.0`. Could you pick it up again and push it over the finish line?

**Comment by purna135** [comment]
Hi @michaelosthege sorry for the delay; I was preoccupied with my exam, which has now been done.
I will complete this RP in few days.

File: pymc/variational/opvi.py [hunk 4]
```diff
@@ -957,9 +957,9 @@ def __init_group__(self, group):
         if not group:
             raise GroupError("Got empty group")
         if self.local:
-            raise NotImplementedInference("Local inferene aka AEVB is not supported in v4")
+            warnings.warn("Local inferene aka AEVB is deprecated.", DeprecationWarning)
```
Comment (twiecki) [inline pymc/variational/opvi.py]: We should just delete this code, rather than deprecate.

**Comment by purna135** [comment]
Hi @twiecki and @ferrine could you please have a look if this is the right way to go?

**Review by ferrine — APPROVED** [review]
Looks good to me, can be merged once tests pass

**Comment by purna135** [comment]
Thank you very much, @ferrine. Could you kindly advise me on how to pass these failed tests?
In the doc string, I realised I left some references to `batched` and `local`, which I will remove in my next commit.

**Comment by ferrine** [comment]
@purna135, the tests diff was collapsed. I think I missed something here. I have to go through it once more

File: pymc/tests/test_variational_inference.py [hunk 5]
```diff
@@ -98,89 +87,6 @@ def three_var_model():
...
-            },
-        ),
-        (
-            pytest.raises(TypeError, match="No approximation is specified"),
-            {MeanFieldGroup: ["one", "two"]},
-        ),
-        (not_raises(), {MeanFieldGroup: ["one"], FullRankGroup: ["two", "three"]}),
-    ],
-)
-@ignore_not_implemented_inference
```
Comment (ferrine) [inline pymc/tests/test_variational_inference.py:125]: the `ignore_not_implemented_inference` hook was used to ignore tests that fail due to the not supported functionality. 

`NotImplementedInference` is an explicit error not to confuse failures with other "bugs". Most of the tests that are deleted so far are valid tests. 

The commit that deletes all these tests should be reverted. Or, if not easy, you can use the original file that is present in master.

To summarise. 

1. The only obsolete thing in the test file is `ignore_not_implemented_inference`. 
2. Once it is removed, all tests should pass and no `NotImplementedInference` is raised.
3. The only tests to remove are that fail because of the removed functionality.

**Review by ferrine — CHANGES_REQUESTED** [review]
I missed the collapsed diff of tests. Most of the deleted tests are valid. I've left a comment above with the explanation.

**Comment by ferrine** [comment]
We did a quick video chat with @purna135 to avoid further confusion. Additional thing to revisit was the documentation. That missed my eye as well.

**Comment by purna135** [comment]
Thank you so much for your time and help @ferrine : )

**Comment by purna135** [comment]
Hello, @ferrine. I deleted several flow-related tests and refactored others, but there are still 10 tests that I am unable to refactor and have commented so that other tests can be checked. Could you please assist me in refactoring the tests I commented? 
Many thanks
